src/containerapp-preview/azext_containerapp_preview/_clients.py

```diff
@@ -0,0 +1,915 @@
+
+import json
+import sys
+import time
+
+from azure.cli.core.azclierror import AzureResponseError, ResourceNotFoundError
+from azure.cli.core.util import send_raw_request
+from azure.cli.core.commands.client_factory import get_subscription_id
+from knack.log import get_logger
+
+from ._utils import (_get_azext_module, GA_CONTAINERAPP_EXTENSION_NAME)
+
+logger = get_logger(__name__)
+
+PREVIEW_API_VERSION = "2023-04-01-preview"
+CURRENT_API_VERSION = "2023-04-01-preview"
+POLLING_TIMEOUT = 600  # how many seconds before exiting
+POLLING_SECONDS = 2  # how many seconds between requests
+POLLING_TIMEOUT_FOR_MANAGED_CERTIFICATE = 1500  # how many seconds before exiting
+POLLING_INTERVAL_FOR_MANAGED_CERTIFICATE = 4  # how many seconds between requests
+HEADER_AZURE_ASYNC_OPERATION = "azure-asyncoperation"
+HEADER_LOCATION = "location"
+
+
+def poll_status(cmd, request_url):
+    azext_client = _get_azext_module(
+        GA_CONTAINERAPP_EXTENSION_NAME, "azext_containerapp._clients")
+    return azext_client.poll_status(cmd, request_url)
+
+
+def poll_results(cmd, request_url):
+    azext_client = _get_azext_module(
+        GA_CONTAINERAPP_EXTENSION_NAME, "azext_containerapp._clients")
+    return azext_client.poll_results(cmd, request_url)
+
+
+class PollingAnimation():
+    def __init__(self):
+        self.tickers = ["/", "|", "\\", "-", "/", "|", "\\", "-"]
+        self.currTicker = 0
+
+    def tick(self):
+        sys.stderr.write('\r')
+        sys.stderr.write(self.tickers[self.currTicker] + " Running ..")
+        sys.stderr.flush()
+        self.currTicker += 1
+        self.currTicker = self.currTicker % len(self.tickers)
+
+    def flush(self):
+        sys.stderr.flush()
+        sys.stderr.write("\r\033[K")
+
+
+class ContainerAppClient():
```

Comment:
duplicate with GA ext?

Reply:
It is copy from GA ext, need duplicate Client for preview.
It will have different api-version or request Url in future.